contracts/VestingManager.sol

```diff
@@ -0,0 +1,212 @@
+
+contract VestingManager is Ownable {
+    using SafeERC20 for IERC20;
+
+    IERC20 public immutable oidToken;
+    uint256 public referenceDate = 0;
+    uint32 public constant QUARTER = 13 weeks;
+    uint8 public schemaId = 0;
+
+    struct VestingSchema {
+        uint8 lockup; // Number of quarters during which the holders tokens are locked and not accountable for vesting
+        uint16 vesting; // % of total tokens released every quarter with two decimals of resolution (i.e. 1256 -> 12,56%)
+    }
+
+    struct Holding {
+        uint256 totalVestedAmount;
+        uint256 releasedAmount;
+        uint8 vestingSchema;
+    }
+
+    /* Mappings */
+    mapping(uint8 => VestingSchema) public VestingSchemas;
+    mapping(address => Holding[]) public Holdings;
+
+    /* Events */
+    event Deposit(address to, uint256 amount, uint8 schemaId);
+    event Withdraw(address to, uint256 amount);
+    event SchemaCreated(uint8 schemaId, uint8 lockup, uint16 vesting);
+    event EthDeposit(address sender, uint256 value, uint256 balance);
+
+    /* Modifiers */
+    modifier referenceInitiated() {
+        require(referenceDate != 0, 'Reference date not initialized');
+        _;
+    }
+
+    constructor(IERC20 oidToken_) {
+        require(address(oidToken_) != address(0), 'Token address cannot be addr(0)');
+        oidToken = oidToken_;
+    }
+
+    /* Admin functions */
+
+    /**
+     * @dev Sets the reference date for all lockups in the contract.
+     * @param _referenceDate period in quarters that tokens remain blocked
+     */
+    function setReferenceDate(uint256 _referenceDate) external onlyOwner {
+        require(referenceDate == 0, 'Reference date has already been initialized');
+        require(_referenceDate != 0, 'Cannot set reference date to 0');
+        referenceDate = _referenceDate;
+    }
+
+    /**
+     * @dev Default withdraw function in case someone deposits ETH erroneously.
+     */
+    function withdraw() external onlyOwner {
+        uint256 amount = address(this).balance;
+
+        (bool success, ) = owner().call{value: amount}('');
+        require(success, 'Failed to send Ether');
+    }
+
+    /**
+     * @dev Withdraws remaining unlocked tokens across offerings.
+     * @param _lockup period in quarters that tokens remain blocked
+     * @param _vesting percentage of total tokens released each quarter
+     * @return created schema ID
+     */
+    function createVestingSchema(uint8 _lockup, uint16 _vesting) external onlyOwner returns (uint8) {
+        // Validate that we are not overflowing uint8
+        uint8 newSchemaId = schemaId;
+        schemaId++;
+        require(_vesting > 0 && _vesting <= 10000, 'Vesting % should be withing 0 and 10000 (2 decimal floating point)');
+        // Store new schema and emit event
+        VestingSchemas[newSchemaId] = VestingSchema(_lockup, _vesting);
+        emit SchemaCreated(newSchemaId, _lockup, _vesting);
+
+        return newSchemaId;
```

Comment:
do we need a return on this function ? if yes, why?

Reply:
Afaik the return was a requirement. Can facilitate new deposits by getting the ID directly after vesting schema creation instead of fetch schemas > create > fetch again to obtain the ID

Reply:
if there is an utility to it, i don't see any problem to keep it indeed, was just wondering the usefulness as we have the ID in the event

Reply:
True, forgot about the event, the only use I can think of is if this function is used internally in the contract. Unless we plan on extending the vestingManager I will remove the return

Reply:
i don't think we plan to extend it to other smart contracts no, so the event is enough indeed :)